### PR TITLE
[Data] Improve stability of Parquet metadata prefetch task

### PR DESCRIPTION
**Pull request by c21**
## Why are these changes needed?

This PR is a fix for Parquet metadata prefetching task, when reading a large amount of Parquet files on S3 (>50k). Before this PR, the Parquet prefetch metadata task is running on head node (w/ `DEFAULT` scheduling strategy), and not retry on S3 transient exception. So it can fail very quickly because it launches too many request from same node, and throttled by S3.

This PR does 3 things:
* Fix scheduling strategy to use `SPREAD` same as read task, to spread out metadata prefetch task across cluster. This avoids hit S3 w/ too many requests from same node.
* Auto-retry on `OSError`, where S3 throws transient error such as `Access Denied`, `Read Timeout`.
* Extract `num_cpus` default value out as a variable. So we can tune the value to control the concurrency of prefetch metadata task for particular workload. Sometime `num_cpus=0.5` does not work well.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by c21** [comment]
It's hard to understand a unit test given it depends on failure from prefetching metadata.

I can run the Parquet metadata prefetching release test, to make sure it still works - https://github.com/ray-project/ray/blob/master/release/release_tests.yaml#L5485 .

File: python/ray/data/datasource/parquet_datasource.py [hunk 1]
```diff
@@ -47,6 +47,15 @@
...
 PARALLELIZE_META_FETCH_THRESHOLD = 24
 
+# The `num_cpus` for each metadata prefetching task.
+# Default to 0.5 instead of 1 because it is cheaper than normal read task.
+NUM_CPUS_FOR_META_FETCH_TASK = 0.5
+
+# The application-level exceptions to retry for metadata prefetching task.
+# Default to retry on `OSError` because AWS S3 would throw this transient
+# error when load is too high.
+RETRY_EXCEPTIONS_FOR_META_FETCH_TASK = [OSError]
```
Comment (scottjlee) [inline python/ray/data/datasource/parquet_datasource.py:57]: should we add this in general for all `FileBasedDatasource` reads? https://github.com/ray-project/ray/blob/97406423d8d6d093f52588006eb55624cbfc27cf/python/ray/data/datasource/file_based_datasource.py#L68

File: python/ray/data/datasource/parquet_datasource.py [hunk 2]
```diff
@@ -47,6 +47,15 @@
 FRAGMENTS_PER_META_FETCH = 6
 PARALLELIZE_META_FETCH_THRESHOLD = 24
 
+# The `num_cpus` for each metadata prefetching task.
+# Default to 0.5 instead of 1 because it is cheaper than normal read task.
+NUM_CPUS_FOR_META_FETCH_TASK = 0.5
```
Comment (scottjlee) [inline python/ray/data/datasource/parquet_datasource.py:52]: so currently, there is no configurable way (e.g. through `DataContext`) to modify this right?

**Review by scottjlee — APPROVED** [review]

File: python/ray/data/datasource/parquet_datasource.py (hunk 1 above)
Comment (c21) [reply python/ray/data/datasource/parquet_datasource.py:57]: Only Parquet has this metadata prefetching. So we can add it first here. For all file-based datasources, looks like the expoential backoff retry is working.

File: python/ray/data/datasource/parquet_datasource.py (hunk 2 above)
Comment (c21) [reply python/ray/data/datasource/parquet_datasource.py:52]: yes, I am hesitating to add to `DataContext` (which exposed to users). If we see this is a value needed by users, we can expose it later. For now, let's start w/ a varilable, so at least we (internal developer) can change it.

File: python/ray/data/tests/test_parquet.py [hunk 3]
```diff
@@ -206,7 +206,10 @@ def test_parquet_read_meta_provider(ray_start_regular_shared, fs, data_path):
     pq.write_table(table, path2, filesystem=fs)
 
     class TestMetadataProvider(DefaultParquetMetadataProvider):
-        def prefetch_file_metadata(self, fragments):
+        def prefetch_file_metadata(self, fragments, **ray_remote_args):
+            assert ray_remote_args["num_cpus"] == 0.5
```
Comment (raulchen) [inline python/ray/data/tests/test_parquet.py:210]:
```suggestion
            assert ray_remote_args["num_cpus"] == NUM_CPUS_FOR_META_FETCH_TASK
```

**Review by raulchen — APPROVED** [review]

File: python/ray/data/tests/test_parquet.py (hunk 3 above)
Comment (c21) [reply python/ray/data/tests/test_parquet.py]: updated.

**Comment by c21** [comment]
Reran the Parquet metadata release test, no regression - https://buildkite.com/ray-project/release/builds/4714#018c89f0-5bde-4011-835d-90d244d2c3f0 .

**Pull request merged**